File: plugin/ios/AnylineSDKPlugin.m

#import <React/RCTEventEmitter.h>
#import <React/RCTBridge.h>
#import "AnylineSDKPlugin.h"
#import "NativeView/NativeViewRegistry.h"
#import "RCTUtils.h"

// Static instance to retain the ALWrapperSessionProvider instance
static ALWrapperSessionProvider *_wrapperSessionProvider;
// We're creating a static variable to store the last license used to initialize the SDK
static NSString *license;

static NSString *scanViewConfigPathString;

@interface AnylineSDKPlugin()

@property (nonatomic, copy) NSString *callbackId;

@property (nonatomic, copy) NSString *config;

@property (nonatomic, copy) NSString *returnMethod;

@property (nonatomic, strong) RCTPromiseResolveBlock wrapperSessionSdkInitializationResponsePromiseResolve;
@property (nonatomic, strong) RCTPromiseRejectBlock wrapperSessionSdkInitializationResponsePromiseReject;

@property (nonatomic, strong) RCTPromiseResolveBlock wrapperSessionScanResponsePromiseResolve;
@property (nonatomic, strong) RCTPromiseRejectBlock wrapperSessionScanResponsePromiseReject;

@property (nonatomic, strong) RCTResponseSenderBlock scanResponseResultCallback;
@property (nonatomic, strong) RCTResponseSenderBlock scanResponseErrorCallback;

@property (nonatomic, strong) RCTPromiseResolveBlock wrapperSessionExportCachedEventsPromiseResolve;
@property (nonatomic, strong) RCTPromiseRejectBlock wrapperSessionExportCachedEventsPromiseReject;

@property (nonatomic, strong) RCTResponseSenderBlock reportCorrectedResultResponseCallback;
@property (nonatomic, strong) RCTResponseSenderBlock reportCorrectedResultErrorCallback;

@end


@implementation AnylineSDKPlugin {
    NSMutableArray<NSString *> *supportedRTCEventsArray;
}

- (instancetype)init {
    self = [super initWithDisabledObservation];
    if (self) {
        if (!_wrapperSessionProvider) {
            // Initialize the wrapperSessionProvider static variable
            _wrapperSessionProvider = [[ALWrapperSessionProvider alloc] init];
        }
        supportedRTCEventsArray = [NSMutableArray array];
    }
    return self;
}

+ (BOOL)isInitializedWithLicenseKey:(NSString *)requestedLicense {
    ALWrapperSessionSDKInitializationResponse *currentSdkInitializationResponse = [ALWrapperSessionProvider getCurrentSdkInitializationResponse];
    return (license
            && [currentSdkInitializationResponse.initialized isEqualToNumber:@YES]
            && [license isEqualToString:requestedLicense]);
}

RCT_EXPORT_MODULE();

- (UIView *)view {
    return [[UIView alloc] initWithFrame:CGRectMake(0, 0, 0, 0)];
}

RCT_EXPORT_METHOD(getSDKVersion:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    resolve(AnylineSDK.versionNumber);
}

RCT_EXPORT_METHOD(setupWrapperSession:(NSString *)pluginVersion) {
    [self setupWrapperSessionWithPluginVersion:pluginVersion];
}

RCT_EXPORT_METHOD(setViewConfigsPath:(NSString *)path) {
    scanViewConfigPathString = path;
}

RCT_EXPORT_METHOD(setupAnylineSDK:(NSString *)licenseKey
        resolver:(RCTPromiseResolveBlock)resolve
        rejecter:(RCTPromiseRejectBlock)reject) {
    [self setupAnylineSDKWithCacheConfig:licenseKey enableOfflineCache:NO resolver:resolve rejecter:reject];
}

RCT_EXPORT_METHOD(setupAnylineSDKWithCacheConfig:(NSString *)licenseKey
                  enableOfflineCache:(BOOL)enableOfflineCache
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {

    self.wrapperSessionSdkInitializationResponsePromiseResolve = resolve;
    self.wrapperSessionSdkInitializationResponsePromiseReject = reject;
    self.returnMethod = @"promise";

    [self initSdkWithLicenseKey:licenseKey sdkAssetsFolder:nil enableOfflineCache:enableOfflineCache];
}

RCT_EXPORT_METHOD(isInitialized:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    ALWrapperSessionSDKInitializationResponse *currentSdkInitializationResponse = [ALWrapperSessionProvider getCurrentSdkInitializationResponse];
    BOOL isInitialized = ([currentSdkInitializationResponse.initialized isEqualToNumber:@YES]);
    resolve(@(isInitialized));
}

RCT_EXPORT_METHOD(licenseKeyExpiryDate:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    ALWrapperSessionSDKInitializationResponse *currentSdkInitializationResponse = [ALWrapperSessionProvider getCurrentSdkInitializationResponse];
    if ([currentSdkInitializationResponse.initialized isEqualToNumber:@YES]) {
        ALWrapperSessionSDKInitializationResponseInitialized *sdkInitializationResponseInitialized = currentSdkInitializationResponse.succeedInfo;
        if (sdkInitializationResponseInitialized) {
            resolve(sdkInitializationResponseInitialized.expiryDate);
            return;
        }
    }
    // not `reject`, as this will require the method to be used in a try-catch
    resolve(nil);
    return;
}

RCT_EXPORT_METHOD(setDefaultScanStartPlatformOptions:(nullable NSString *)scanStartPlatformOptionsString
                                            resolver:(RCTPromiseResolveBlock)resolve
                                            rejecter:(RCTPromiseRejectBlock)reject) {
    @try {
        [ALWrapperSessionProvider setDefaultScanStartPlatformOptionsWithString:scanStartPlatformOptionsString];
        resolve(@"");
    }
    @catch (NSException *exception) {
        reject(@"ANYLINE_ERROR", exception.reason, nil);
    }
}

// Deprecated
RCT_EXPORT_METHOD(setupScanViewWithConfigJson:(NSString *)config scanMode:(NSString *)scanMode onResultCallback:(RCTResponseSenderBlock)onResult onErrorCallback:(RCTResponseSenderBlock)onError) {
    [self setupWithInitializationParameters:nil config:config scanMode:scanMode onResultCallback:onResult onErrorCallback:onError];
}

RCT_EXPORT_METHOD(setup:(NSString *)config scanMode:(NSString *)scanMode onResultCallback:(RCTResponseSenderBlock)onResult onErrorCallback:(RCTResponseSenderBlock)onError) {
    [self setupWithInitializationParameters:nil config:config scanMode:scanMode onResultCallback:onResult onErrorCallback:onError];
}

RCT_EXPORT_METHOD(setupWithInitializationParameters:(NSString * _Nullable)initializationParametersStr config:(NSString *)config scanMode:(NSString *)scanMode onResultCallback:(RCTResponseSenderBlock)onResult onErrorCallback:(RCTResponseSenderBlock)onError) {
    self.scanResponseResultCallback = onResult;
    self.scanResponseErrorCallback = onError;
    self.returnMethod = @"callback";

    [self requestScanStartWithScanViewConfigContent:config
             scanViewInitializationParametersString:initializationParametersStr
                                 scanViewConfigPath:scanViewConfigPathString
                           scanCallbackConfigString:nil];
}

RCT_EXPORT_METHOD(setupPromise:(NSString *)config scanMode:(NSString *)scanMode resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [self setupPromiseWithInitializationParameters:nil config:config scanMode:scanMode resolver:resolve rejecter:reject];
}

RCT_EXPORT_METHOD(setupPromiseWithInitializationParameters:(NSString * _Nullable)initializationParametersStr
                  config:(NSString *)configStr
                  scanMode:(NSString *)scanMode
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [self setupPromiseWithInitializationParametersAndScanCallbackConfig:initializationParametersStr
                                                                 config:configStr
                                                               scanMode:scanMode
                                               scanCallbackConfigString:nil
                                                               resolver:resolve
                                                               rejecter:reject];
}

RCT_EXPORT_METHOD(setupPromiseWithScanCallbackConfig:(NSString *)configStr
                  scanMode:(NSString *)scanMode
                  scanCallbackConfigString:(NSString *)scanCallbackConfigString
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [self setupPromiseWithInitializationParametersAndScanCallbackConfig:nil
                                                                 config:configStr
                                                               scanMode:scanMode
                                               scanCallbackConfigString:scanCallbackConfigString
                                                               resolver:resolve
                                                               rejecter:reject];
}

RCT_EXPORT_METHOD(setupPromiseWithInitializationParametersAndScanCallbackConfig:(NSString * _Nullable)initializationParametersStr
                  config:(NSString *)configStr
                  scanMode:(NSString *)scanMode
                  scanCallbackConfigString:(NSString * _Nullable)scanCallbackConfigString
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    self.wrapperSessionScanResponsePromiseResolve = resolve;
    self.wrapperSessionScanResponsePromiseReject = reject;
    self.returnMethod = @"promise";
    [self requestScanStartWithScanViewConfigContent:configStr
             scanViewInitializationParametersString:initializationParametersStr
                                 scanViewConfigPath:scanViewConfigPathString
                           scanCallbackConfigString:scanCallbackConfigString];
}

RCT_EXPORT_METHOD(trySwitchScan:(NSString *)scanViewConfigContent) {
    [ALWrapperSessionProvider requestScanSwitchWithScanViewConfigContentString:scanViewConfigContent];
}

RCT_EXPORT_METHOD(tryStopScan:(NSString * _Nullable)scanStopRequestParams) {
    [ALWrapperSessionProvider requestScanStopWithScanStopRequestParamsString:scanStopRequestParams];
}

RCT_EXPORT_METHOD(reportCorrectedResult:(NSString *)blobKey
                        correctedResult:(NSString *)correctedResult
                       onResultCallback:(RCTResponseSenderBlock)onResult
                        onErrorCallback:(RCTResponseSenderBlock)onError) {
    self.reportCorrectedResultResponseCallback = onResult;
    self.reportCorrectedResultErrorCallback = onError;

    [self requestUCRReportWithBlobKey:blobKey correctedResult:correctedResult];
}


RCT_EXPORT_METHOD(exportCachedEvents:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    self.wrapperSessionExportCachedEventsPromiseResolve = resolve;
    self.wrapperSessionExportCachedEventsPromiseReject = reject;
    [self exportCachedEvents];
}


- (NSArray<NSString *> *)supportedEvents {
  return supportedRTCEventsArray;
}

- (void)sendEvent:(NSString *)eventName
           params:(id _Nullable) params {

    if (![supportedRTCEventsArray containsObject:eventName]) {
        [supportedRTCEventsArray addObject:eventName];
    }

    [self sendEventWithName:eventName body:params];
}

-(NSString *)bundleRootPath {
    NSString *rootPath = [[NSBundle mainBundle] bundlePath];
    return rootPath;
}

-(NSString *)bundlePathFromScanViewConfigPath:(NSString * _Nullable)scanViewConfigPath {
    //get root folder (ends with ...Runner.app)
    NSString *absoluteScanViewConfigPath = [self bundleRootPath];

    if (scanViewConfigPath) {
        //append scanViewConfigPath
        absoluteScanViewConfigPath = [absoluteScanViewConfigPath stringByAppendingPathComponent:scanViewConfigPath];
    }
    return absoluteScanViewConfigPath;
}

- (void)setupWrapperSessionWithPluginVersion:(NSString *)pluginVersion {
    // Setup wrapper session with this view controller as delegate
    ALWrapperConfig *wrapperConfig = [ALWrapperConfig reactNative:pluginVersion];
    [ALWrapperSessionProvider setupWrapperSessionWithWrapperInfo:wrapperConfig
                                            wrapperSessionClient:self];
}

- (void)initSdkWithLicenseKey:(NSString *)sdkLicenseKey
              sdkAssetsFolder: (NSString *)sdkAssetsFolder
           enableOfflineCache: (BOOL)enableOfflineCache {
    NSDictionary *wrapperSessionSdkInitializationRequestJson = [ALLegacyPluginHelper
            sdkInitializationRequestJsonWithLicenseKey:sdkLicenseKey
                                    enableOfflineCache:enableOfflineCache
                                       assetPathPrefix:sdkAssetsFolder];

    [ALWrapperSessionProvider
            requestSdkInitializationWithInitializationRequestParamsString:[wrapperSessionSdkInitializationRequestJson asJSONString]];
}

- (void)requestScanStartWithScanViewConfigContent:(NSString *)scanViewConfigContent
           scanViewInitializationParametersString:(NSString * _Nullable)scanViewInitializationParametersString
                               scanViewConfigPath:(NSString * _Nullable)scanViewConfigPath
                         scanCallbackConfigString:(NSString * _Nullable)scanCallbackConfigString {
    NSError *error;
    BOOL shouldReturnImages = true;

    ALWrapperSessionScanStartRequest *wrapperSessionScanStartRequest = [ALLegacyPluginHelper
                                        scanStartRequestWithScanViewConfigContentString:scanViewConfigContent
                                                 scanViewInitializationParametersString:scanViewInitializationParametersString
                                                                     scanViewConfigPath:[self bundlePathFromScanViewConfigPath:scanViewConfigPath]
                                                         scanResultCallbackConfigString:scanCallbackConfigString
                                                                     shouldReturnImages:shouldReturnImages
                                                                                  error:&error];
    if (error) {
        [self returnError:error];
    } else {
        NSDictionary *wrapperSessionScanStartRequestDict = [wrapperSessionScanStartRequest toJSONDictionary];
        [ALWrapperSessionProvider requestScanStartWithScanStartRequestParamsString:[wrapperSessionScanStartRequestDict asJSONString]];
    }
}

- (void)requestUCRReportWithBlobKey:(NSString *)blobKey
                    correctedResult:(NSString *)correctedResult {

    ALWrapperSessionUCRReportRequest *wrapperSessionSdkInitializationRequest = [ALLegacyPluginHelper ucrReportRequestWithBlobKey:blobKey
                                                                                  correctedResult:correctedResult];
    [ALWrapperSessionProvider requestUCRReport:wrapperSessionSdkInitializationRequest];
}

-(void)exportCachedEvents {
    [ALWrapperSessionProvider requestExportCachedEvents];
}

- (void)returnCallback:(id _Nullable)callbackObj andError:(NSError * _Nullable)error {
    NSString *resultStr;
    NSError *errorObj;
    if (error) {
            [self returnError:error];
    } else if ([NSJSONSerialization isValidJSONObject:callbackObj]) {
        resultStr = [(NSDictionary *)callbackObj toJSONStringPretty:YES error:&errorObj];
        if (errorObj) {
            [self returnError:errorObj];
        }
        [self returnSuccess:resultStr];
    } else {
        resultStr = @"callback object should be of JSON type";
        [self returnError:[NSError errorWithDomain:@"ALReactDomain" code:100 userInfo:@{NSLocalizedDescriptionKey: resultStr}]];
    }
}

- (void)returnSuccess:(NSString *)result {
    if ([self.returnMethod isEqualToString:@"callback"]) {
        self.scanResponseResultCallback(@[result]);
    } else if ([self.returnMethod isEqualToString:@"promise"]) {
        if (self.wrapperSessionScanResponsePromiseResolve) {
            self.wrapperSessionScanResponsePromiseResolve(result);
        }
        self.wrapperSessionScanResponsePromiseResolve = nil;
    }
}

- (void)returnError:(NSError *)error {
    if ([self.returnMethod isEqualToString:@"callback"]) {
        self.scanResponseErrorCallback(@[error]);
    } else if ([self.returnMethod isEqualToString:@"promise"]) {
        if (self.wrapperSessionScanResponsePromiseReject) {
            self.wrapperSessionScanResponsePromiseReject(@"ANYLINE_ERROR", error.localizedDescription, error);
        }
        self.wrapperSessionScanResponsePromiseReject = nil;
    }
}

#pragma mark - ALWrapperSessionClientDelegate

- (nullable UIViewController *)getTopViewController {
    UIViewController *topViewController = RCTPresentedViewController();
    if (!topViewController) {
        topViewController = [[UIApplication sharedApplication] keyWindow].rootViewController;
        return topViewController;
    }
    return nil;
}

- (nullable UIView *)getContainerView {
    return [[NativeViewRegistry shared] getLastOrNull];
}

- (void)onSdkInitializationResponse:(nonnull ALWrapperSessionSDKInitializationResponse *)initializationResponse {
    if ([initializationResponse.initialized isEqualToNumber:@YES]) {
        self.wrapperSessionSdkInitializationResponsePromiseResolve(@"success");
    } else {
        self.wrapperSessionSdkInitializationResponsePromiseReject(@"ANYLINE_ERROR", [[initializationResponse toJSONDictionary] asJSONString], nil);
    }
}

- (void)onScanResults:(nonnull ALWrapperSessionScanResultsResponse *)scanResultsResponse {
    ALWrapperSessionScanResultConfig *scanResultConfig = scanResultsResponse.scanResultConfig;
    NSArray<ALExportedScanResult *> *exportedScanResultsArray =  (NSArray<ALExportedScanResult *> *) scanResultsResponse.exportedScanResults;
    ALWrapperSessionScanResultExtraInfo *scanResultExtraInfo = scanResultsResponse.scanResultExtraInfo;

    NSError *error;
    NSString *resultsWithImagePathString = [ALLegacyPluginHelper scanResultsWithImagePathFromExportedScanResults:exportedScanResultsArray
                                                                                                          viewPluginType:scanResultExtraInfo.viewPluginType
                                                                                                                   error:&error];

    if (scanResultConfig.callbackConfig && scanResultConfig.callbackConfig.onResultEventName) {
        [self sendEvent:scanResultConfig.callbackConfig.onResultEventName params:resultsWithImagePathString];
    } else {
        [self returnSuccess:resultsWithImagePathString];
    }
}

- (void)onScanResponse:(nonnull ALWrapperSessionScanResponse *)scanResponse {
    if (scanResponse.status == ALWrapperSessionScanResponseStatus.scanSucceeded) {
        ALWrapperSessionScanResultConfig *scanResultConfig = scanResponse.scanResultConfig;
        if (scanResultConfig.callbackConfig && scanResultConfig.callbackConfig.onResultEventName) {
            [self returnSuccess:@""];
        }
    } else if (scanResponse.status == ALWrapperSessionScanResponseStatus.scanFailed) {
        [self returnError:[NSError errorWithDomain:@"ALReactDomain" code:100 userInfo:@{NSLocalizedDescriptionKey: scanResponse.failInfo.lastError}]];
    } else if (scanResponse.status == ALWrapperSessionScanResponseStatus.scanAborted) {
        [self returnError:[NSError errorWithDomain:@"ALReactDomain" code:100 userInfo:@{NSLocalizedDescriptionKey: @"Canceled"}]];
    }
}

- (void)onUIElementClicked:(nonnull ALWrapperSessionScanResultConfig *)scanResultConfig
        uiFeedbackElementConfig:(nonnull ALUIFeedbackElementConfig *)uiFeedbackElementConfig {
    if (scanResultConfig.callbackConfig && scanResultConfig.callbackConfig.onUIElementClickedEventName) {
        [self sendEvent:scanResultConfig.callbackConfig.onUIElementClickedEventName
                 params:uiFeedbackElementConfig];
    }
}

- (void)onUCRReportResponse:(nonnull ALWrapperSessionUCRReportResponse *)ucrReportResponse {
    if (ucrReportResponse.status == ALWrapperSessionUCRReportResponseStatus.ucrReportSucceeded) {
        ALWrapperSessionUCRReportResponseSucceed *ucrReportSucceed = ucrReportResponse.succeedInfo;
        if (self.reportCorrectedResultResponseCallback) {
            self.reportCorrectedResultResponseCallback(@[ucrReportSucceed.message]);
        }
    } else {
        ALWrapperSessionUCRReportResponseFail *ucrReportFail = ucrReportResponse.failInfo;
        if (self.reportCorrectedResultErrorCallback) {
            self.reportCorrectedResultErrorCallback(@[ucrReportFail.lastError]);
        }
    }
}

- (void)onExportCachedEventsResponse:(nonnull ALWrapperSessionExportCachedEventsResponse *)exportCachedEventsResponse {
    if (exportCachedEventsResponse.status == ALWrapperSessionExportCachedEventsResponseStatus.exportSucceeded) {
        ALWrapperSessionExportCachedEventsResponseSucceed *exportCachedEventsSucceed = exportCachedEventsResponse.succeedInfo;
        if (self.wrapperSessionExportCachedEventsPromiseResolve) {
            self.wrapperSessionExportCachedEventsPromiseResolve(@[exportCachedEventsSucceed.exportedFile]);
        }
    } else {
        ALWrapperSessionExportCachedEventsResponseFail *exportCachedEventsFail = exportCachedEventsResponse.failInfo;
        if (self.wrapperSessionExportCachedEventsPromiseReject) {
            self.wrapperSessionExportCachedEventsPromiseReject(@"ANYLINE_ERROR", exportCachedEventsFail.lastError, nil);
        }
    }
}

@end

File: plugin/android/src/main/java/com/anyline/reactnative/AnylineSDKPlugin.java
package com.anyline.reactnative;

/**
 * Created by jonesBoi on 02.12.16.
 */

import android.content.Context;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;

import java.util.List;

import io.anyline.plugin.config.UIFeedbackElementConfig;
import io.anyline.plugin.result.ExportedScanResult;
import io.anyline.wrapper.config.WrapperSessionExportCachedEventsResponse;
import io.anyline.wrapper.config.WrapperSessionExportCachedEventsResponseSucceed;
import io.anyline.wrapper.config.WrapperSessionExportCachedEventsResponseFail;
import io.anyline.wrapper.config.WrapperSessionScanResultExtraInfo;
import io.anyline.wrapper.config.WrapperSessionScanResultsResponse;
import io.anyline.wrapper.config.WrapperSessionScanStartRequest;
import io.anyline.wrapper.config.WrapperSessionScanResponse;
import io.anyline.wrapper.config.WrapperSessionScanResultConfig;
import io.anyline.wrapper.config.WrapperSessionSdkInitializationResponse;
import io.anyline.wrapper.config.WrapperSessionSdkInitializationResponseInitialized;
import io.anyline.wrapper.config.WrapperSessionUCRReportRequest;
import io.anyline.wrapper.config.WrapperSessionUCRReportResponse;
import io.anyline2.WrapperInfo;
import io.anyline2.di.context.ContextProvider;
import io.anyline2.sdk.extension.UIFeedbackElementConfigExtensionKt;
import io.anyline2.wrapper.WrapperSessionClientInterface;
import io.anyline2.wrapper.WrapperSessionProvider;
import io.anyline2.wrapper.extensions.WrapperSessionScanStartRequestExtensionKt;
import io.anyline2.wrapper.extensions.WrapperSessionSdkInitializationResponseExtensionKt;
import io.anyline2.wrapper.extensions.WrapperSessionUCRReportRequestExtensionKt;
import io.anyline2.wrapper.legacy.LegacyPluginHelper;

class AnylineSDKPlugin extends ReactContextBaseJavaModule
        implements WrapperSessionClientInterface, ResultReporter.OnResultListener {

    // We're creating a static variable to retain the WrapperSessionProvider instance in order to prevent crashes due to garbage collection cleaning up the SDK.
    protected static final WrapperSessionProvider wrapperSessionProvider = WrapperSessionProvider.INSTANCE;
    // We're creating a static variable to store the last license used to initialize the SDK
    private static String license;

    private static boolean isInitializedWithLicenseKey(String requestedLicense) {
        return (license != null
                && WrapperSessionProvider.getCurrentSdkInitializationResponse().getInitialized() == Boolean.TRUE
                && license.equals(requestedLicense));
    }

    public static final String REACT_CLASS = "AnylineSDKPlugin";
    private static final String E_ERROR = "E_ERROR";

    private final ReactApplicationContext reactContext;

    private String returnMethod;

    private Promise wrapperSessionSdkInitializationResponsePromise = null;

    private Promise wrapperSessionScanResponsePromise = null;
    private Callback scanResponseResultCallback = null;
    private Callback scanResponseErrorCallback = null;

    private Promise wrapperSessionExportCachedEventsPromise = null;

    private Callback reportCorrectedResultResponseCallback = null;
    private Callback reportCorrectedResultErrorCallback = null;

    AnylineSDKPlugin(ReactApplicationContext context) {
        super(context);
        this.reactContext = context;
        ContextProvider.setInstance(context);
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @ReactMethod
    public void getSDKVersion(final Promise promise) {
        promise.resolve(at.nineyards.anyline.BuildConfig.VERSION_NAME);
    }

    @ReactMethod
    protected void setupWrapperSession(final String pluginVersion) {
        WrapperInfo wrapperInfo = new WrapperInfo(WrapperInfo.WrapperType.ReactNative, pluginVersion);
        WrapperSessionProvider.setupWrapperSession(wrapperInfo,this);
    }

    @ReactMethod
    public void licenseKeyExpiryDate(final Promise promise) {
        if (WrapperSessionProvider.getCurrentSdkInitializationResponse().getInitialized() == Boolean.TRUE) {
            WrapperSessionSdkInitializationResponseInitialized sdkInitializationResponseInitialized
                    = WrapperSessionProvider.getCurrentSdkInitializationResponse().getSucceedInfo();
            if (sdkInitializationResponseInitialized != null) {
                promise.resolve(sdkInitializationResponseInitialized.getExpiryDate());
                return;
            }
        }
        promise.reject(E_ERROR, "Anyline SDK was not initialized");
    }

    @ReactMethod
    public void isInitialized(final Promise promise) {
        promise.resolve(WrapperSessionProvider.getCurrentSdkInitializationResponse().getInitialized());
    }

    @ReactMethod
    @Deprecated
    public void setupScanViewWithConfigJson(String config, String scanMode, Callback onResultReact, Callback onErrorReact) {
        scanResponseResultCallback = onResultReact;
        scanResponseErrorCallback = onErrorReact;
        this.returnMethod = "callback";

        routeScanMode(config, null, null, null);
    }

    @ReactMethod
    public void setupAnylineSDK(String license, final Promise promise) {
        setupAnylineSDKWithCacheConfig(license, false, promise);
    }

    @ReactMethod
    public void setupAnylineSDKWithCacheConfig(String license, boolean enableOfflineCache, final Promise promise) {
        if (isInitializedWithLicenseKey(license)) {
            promise.resolve(true);
            return;
        }
        wrapperSessionSdkInitializationResponsePromise = promise;
        initSdkWithCacheConfig(license, enableOfflineCache);
    }

    @ReactMethod
    public void initSdk(String license) {
        initSdkWithCacheConfig(license, false);
    }

    @ReactMethod
    public void initSdkWithCacheConfig(String license, boolean enableOfflineCache) {
        if (isInitializedWithLicenseKey(license)) {
            return;
        }
        AnylineSDKPlugin.license = license;

        JSONObject wrapperSessionSdkInitializationRequestJson =
                LegacyPluginHelper.getWrapperSessionSdkInitializationRequestJson(license, enableOfflineCache, null);

        WrapperSessionProvider.requestSdkInitialization(
                wrapperSessionSdkInitializationRequestJson.toString());
    }

    @ReactMethod
    public void setup(String config, String scanMode, Callback onResultReact, Callback onErrorReact) {
        setupWithInitializationParameters(null, config, scanMode, onResultReact, onErrorReact);
    }
    @ReactMethod
    public void setupWithInitializationParameters(String scanViewInitializationParametersString, String config, String scanMode, Callback onResultReact, Callback onErrorReact) {
        scanResponseResultCallback = onResultReact;
        scanResponseErrorCallback = onErrorReact;
        this.returnMethod = "callback";

        routeScanMode(config, scanViewInitializationParametersString, null, null);
    }
    @ReactMethod
    public void setupPromise(String config, String scanMode, final Promise promise) {
        setupPromiseWithInitializationParameters(null, config, scanMode, promise);
    }
    @ReactMethod
    public void setupPromiseWithScanCallbackConfig(String config, String scanMode, String scanCallbackConfigString, final Promise promise) {
        setupPromiseWithInitializationParametersAndScanCallbackConfig(null, config, scanMode, scanCallbackConfigString, promise);
    }
    @ReactMethod
    public void setupPromiseWithInitializationParameters(String initializationParametersString, String config, String scanMode, final Promise promise) {
        setupPromiseWithInitializationParametersAndScanCallbackConfig(initializationParametersString, config, scanMode, null, promise);
    }
    @ReactMethod
    public void setupPromiseWithInitializationParametersAndScanCallbackConfig(String initializationParametersString, String config, String scanMode, String scanCallbackConfigString, final Promise promise) {
        this.wrapperSessionScanResponsePromise = promise;
        this.returnMethod = "promise";

        routeScanMode(config, initializationParametersString, null, scanCallbackConfigString);
    }

    private void routeScanMode(
            String scanViewConfigContent,
            String scanViewInitializationParametersString,
            String scanViewConfigPath,
            String scanCallbackConfigString) {
        boolean shouldReturnImages = true;

        WrapperSessionScanStartRequest wrapperSessionScanRequest;
        try {
            wrapperSessionScanRequest = LegacyPluginHelper.getWrapperSessionScanStartRequest(
                    this.reactContext,
                    scanViewConfigContent,
                    scanViewInitializationParametersString,
                    scanViewConfigPath,
                    scanCallbackConfigString,
                    shouldReturnImages);
        } catch (Exception e) {
            returnError("Could not parse parameters: " + e.getMessage());
            return;
        }

        JSONObject wrapperSessionScanStartRequestJson
                = WrapperSessionScanStartRequestExtensionKt.toJsonObject(wrapperSessionScanRequest);

        WrapperSessionProvider.requestScanStart(wrapperSessionScanStartRequestJson.toString());

        ResultReporter.setListener(this);
    }

    @ReactMethod
    public void reportCorrectedResult(String blobKey, String correctedResult, Callback onResponseCallback, Callback onErrorCallback) {
        reportCorrectedResultResponseCallback = onResponseCallback;
        reportCorrectedResultErrorCallback = onErrorCallback;
        WrapperSessionUCRReportRequest wrapperSessionUCRReportRequest = LegacyPluginHelper
                .getWrapperSessionUCRReportRequest(blobKey, correctedResult);

        JSONObject wrapperSessionUCRReportRequestJson = WrapperSessionUCRReportRequestExtensionKt
                .toJsonObject(wrapperSessionUCRReportRequest);

        WrapperSessionProvider.requestUCRReport(wrapperSessionUCRReportRequestJson.toString());
    }

    @ReactMethod
    public void exportCachedEvents(final Promise promise) {
        wrapperSessionExportCachedEventsPromise = promise;
        WrapperSessionProvider.requestExportCachedEvents();
    }

    @ReactMethod
    public void tryStopScan(String scanStopRequestParams) {
        WrapperSessionProvider.requestScanStop(scanStopRequestParams);
    }

    @Override
    public void onResult(Object result, boolean isFinalResult) {
        returnSuccess(result.toString());
    }

    @Override
    public void onError(String error) {
        returnError(error);
    }

    @Override
    public void onCancel() {
        returnError("Canceled");
    }

    private void returnError(String error) {
        switch (this.returnMethod) {
            case "callback":
                if (scanResponseErrorCallback != null) {
                    scanResponseErrorCallback.invoke(error);
                    scanResponseErrorCallback = null;
                }
                break;
            case "promise":
                if (wrapperSessionScanResponsePromise != null) {
                    wrapperSessionScanResponsePromise.reject(E_ERROR, error);
                }
                break;
            default:
                break;
        }
    }

    private void returnSuccess(String result) {
        switch (this.returnMethod) {
            case "callback":
                if (scanResponseResultCallback != null) {
                    scanResponseResultCallback.invoke(result);
                    scanResponseResultCallback = null;
                }
                break;
            case "promise":
                if (wrapperSessionScanResponsePromise != null) {
                    wrapperSessionScanResponsePromise.resolve(result);
                }
                break;
            default:
                break;
        }
    }

    private void sendEvent(String eventName, Object params) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }

    @Override
    public @NotNull Context getContext() {
        return this.reactContext;
    }

    @Override
    public void onSdkInitializationResponse(@NotNull WrapperSessionSdkInitializationResponse initializationResponse) {
        JSONObject json =
                WrapperSessionSdkInitializationResponseExtensionKt.toJsonObject(initializationResponse);

        if (wrapperSessionSdkInitializationResponsePromise != null) {
            if (initializationResponse.getInitialized() == Boolean.TRUE) {
                wrapperSessionSdkInitializationResponsePromise.resolve(true);
            } else {
                wrapperSessionSdkInitializationResponsePromise.reject(E_ERROR, json.toString());
            }
        }
    }

    @Override
    public void onScanResults(@NotNull WrapperSessionScanResultsResponse scanResultsResponse) {
        WrapperSessionScanResultConfig scanResultConfig = scanResultsResponse.getScanResultConfig();
        List<ExportedScanResult> scanResultList = scanResultsResponse.getExportedScanResults();
        WrapperSessionScanResultExtraInfo scanResultExtraInfo = scanResultsResponse.getScanResultExtraInfo();
        try {
            String resultsWithImagePathString = LegacyPluginHelper
                    .getScanResultsWithImagePath(scanResultList, scanResultExtraInfo.getViewPluginType());

            if (scanResultConfig.getCallbackConfig() != null
                    &&  scanResultConfig.getCallbackConfig().getOnResultEventName() != null) {
                sendEvent(
                        scanResultConfig.getCallbackConfig().getOnResultEventName(),
                        resultsWithImagePathString);
            } else {
                ResultReporter.onResult(resultsWithImagePathString, true);
            }
        } catch (Exception e) {
            //exception will not be handled here
        }
    }

    @Override
    public void onUIElementClicked(@NonNull WrapperSessionScanResultConfig scanResultConfig,
                                   @NonNull UIFeedbackElementConfig uiFeedbackElementConfig) {
        if (scanResultConfig.getCallbackConfig() != null
                &&  scanResultConfig.getCallbackConfig().getOnUIElementClickedEventName() != null) {
            sendEvent(
                    scanResultConfig.getCallbackConfig().getOnUIElementClickedEventName(),
                    UIFeedbackElementConfigExtensionKt.toJsonObject(uiFeedbackElementConfig).toString());
        }
    }

    @Override
    public void onScanResponse(@NotNull WrapperSessionScanResponse scanResponse) {
        if (scanResponse.getStatus() != null) {
            switch (scanResponse.getStatus()) {
                case SCAN_SUCCEEDED:
                    WrapperSessionScanResultConfig scanResultConfig = scanResponse.getScanResultConfig();
                    if (scanResultConfig.getCallbackConfig() != null
                            &&  scanResultConfig.getCallbackConfig().getOnResultEventName() != null) {
                        ResultReporter.onResult("", true);
                    }
                    break;
                case SCAN_FAILED:
                    ResultReporter.onError(scanResponse.getFailInfo().getLastError());
                    break;
                case SCAN_ABORTED:
                    ResultReporter.onCancel();
                    break;
            }
        }
    }

    @Override
    public void onUCRReportResponse(@NotNull WrapperSessionUCRReportResponse ucrReportResponse) {
        if (ucrReportResponse.getStatus() == WrapperSessionUCRReportResponse.WrapperSessionUCRReportResponseStatus.UCR_REPORT_SUCCEEDED) {
            reportCorrectedResultResponseCallback.invoke(ucrReportResponse.getSucceedInfo().getMessage());
        } else {
            reportCorrectedResultErrorCallback.invoke(LegacyPluginHelper
                    .getWrapperSessionUCRReportResponseFailMessage(ucrReportResponse.getFailInfo()));
        }
        reportCorrectedResultResponseCallback = null;
        reportCorrectedResultErrorCallback = null;
    }

    @Override
    public void onExportCachedEventsResponse(@NotNull WrapperSessionExportCachedEventsResponse exportCachedEventsResponse) {
        if (exportCachedEventsResponse.getStatus() == WrapperSessionExportCachedEventsResponse.WrapperSessionExportCachedEventsResponseStatus.EXPORT_SUCCEEDED) {
            WrapperSessionExportCachedEventsResponseSucceed exportCachedEventsSucceed = exportCachedEventsResponse.getSucceedInfo();
            wrapperSessionExportCachedEventsPromise.resolve(exportCachedEventsSucceed.getExportedFile());
        } else {
            WrapperSessionExportCachedEventsResponseFail exportCachedEventsFail = exportCachedEventsResponse.getFailInfo();
            wrapperSessionExportCachedEventsPromise.reject(E_ERROR, exportCachedEventsFail.getLastError());
        }
        wrapperSessionExportCachedEventsPromise = null;
    }
}